packages/kokkos-kernels/src/sparse/KokkosSparse_spadd.hpp

```diff
@@ -0,0 +1,681 @@
+//
+//               KokkosKernels 0.9: Linear Algebra and Graph Kernels
+//                 Copyright 2017 Sandia Corporation
+//
+// Under the terms of Contract DE-AC04-94AL85000 with Sandia Corporation,
+// the U.S. Government retains certain rights in this software.
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are
+// met:
+//
+// 1. Redistributions of source code must retain the above copyright
+// notice, this list of conditions and the following disclaimer.
+//
+// 2. Redistributions in binary form must reproduce the above copyright
+// notice, this list of conditions and the following disclaimer in the
+// documentation and/or other materials provided with the distribution.
+//
+// 3. Neither the name of the Corporation nor the names of the
+// contributors may be used to endorse or promote products derived from
+// this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY SANDIA CORPORATION "AS IS" AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+// PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL SANDIA CORPORATION OR THE
+// CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+// LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+// NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// Questions? Contact Siva Rajamanickam (srajama@sandia.gov)
+//
+// ************************************************************************
+//@HEADER
+*/
+
+#ifndef _KOKKOS_SPADD_HPP
+#define _KOKKOS_SPADD_HPP
+
+#include "KokkosKernels_Handle.hpp"
+
+namespace KokkosSparse {
+namespace Experimental {
+
+  //Symbolic: count entries in each row in C to produce rowmap
+  //kernel handle has information about whether it is sorted add or not.
+
+  /*
+    Unsorted symbolic algorithm notes:
+      -Only needs to sort and merge indices once, in symbolic (sorting is expensive)
+      -Can't afford to allocate dense Views for indices/values (assume number of columns is very large)
+      -Want numeric() to know exactly where each A/B entry belongs in Ccolinds/Cvalues
+      -To accomplish all of these, symbolic() computes arrays Apos and Bpos (both are type clno_nnz_view_t_,
+        and have same length as a_entries and b_entries respectively)
+        -Apos/Bpos are saved in the handle
+      -Apos and Bpos each contain the final index within C row where the A/B entry belongs
+      -See UnsortedNumericSumFunctor below for the usage of Apos/Bpos
+  */
+
+  //get C rowmap for sorted input
```

Comment:
Trying to mix types between A and B here ? Why are they separate parameters ? Can we have default where types are  same.

Reply:
@srajama1 I can do that. Mehmet suggested by email to have the A/B/C types all specified separately, but I think that was just because he copy-pasted it the function prototypes from spgemm. There will never be a use case for that in Tpetra, and it would make the code look a lot cleaner.

Reply:
Siva, Brian, 
It is not the "data type" (whether it is float/double integer) that is mixed through with this interface. 
Having different types for A/B/C allows us to pass views with different memory spaces. 
We make sure that the data types are same within within our specialization layer.

Reply:
I got that after typing that comment ! I am not sure we need this functionality for spadd though. It is just complexity we would have to maintain for a long time.

Reply:
Actually, as long as the main interface allows different types for A/B/C we do not have to propagate all this types within the code to the functors. We will do the type casting in the specialization layer,  then functors will not need to worry about this. I can help with this.

Reply:
@mndevec Thanks for explaining! My understanding is that all the internal functors should have template parameters scalar_t, size_type, index_type, A_memory_space, B_memory_space, C_memory_space?

And then in add_symbolic/add_numeric we keep the current template parameters, but static_assert that A, B and C scalar/size_type/index_type all match?

Reply:
@brian-kelley 
Firstly, you do not need to worry most of these things. The current way looks good to me. If you push these changes to kokkos-kernels, I can take care of the rest, and point you to these changes. But below are my answers.
 
Yes, for the second question. 

The general interface is as general as possible. You can see the example here:  https://github.com/kokkos/kokkos-kernels/blob/develop/src/sparse/KokkosSparse_spgemm_numeric.hpp

We want this interface to be callable by users without requiring them to change/cast their views. 
The only requirement we have is that "data types should match". This interface will make sure that data types matches for different variants. 

Within the SpGEMM code, there is no operation that depends on the memory space of these views. I think this is almost true for all codes we have there, unless there is some extra specialization one does. So, this interface will cast all memory spaces to unmanaged memory spaces starting from line 167, so that, we do not compile again and again for different memory spaces. 

Than at line 249, I call the internal function. At this point, with casts, you limit the input types, therefore you are free to choose how the internal functions are restricted. For the first part of your question, you can do it freely as you like, as after this entry point you are limiting the input types. 

The internal code I have for spgemm still uses all different templates, as the entry point is implemented after all is implemented in my case :). But still, I use all templates in the internal code, because it allows me to maximize the code reuse internally. 

By the way, I don't know if distance-2 coloring is included in this push. But you do not need to worry about that either. We started integrating your distance-2 coloring code as practice with Will (https://github.com/william76/kokkos-kernels/blob/distance-2-coloring/src/graph/KokkosGraph_D2_GraphColor.hpp). So, you do not need to spend time on that.